blobs/p8z77-m_pro/download_BIOS_clean.sh

```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+# P7 ASUS
+
+function printusage {
+  echo "Usage: $0 -m <me_cleaner>(optional)"
+}
+
+BLOBDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+if [ "$#" -eq 0 ]; then printusage; fi
+
+while getopts ":m:i:" opt; do
+  case $opt in
+    m)
+      if [ -x "$OPTARG" ]; then
+        MECLEAN="$OPTARG"
+      fi
+      ;;
+  esac
+
+done
+
+if [ -z "$MECLEAN" ]; then
+  MECLEAN=`command -v $BLOBDIR/../../build/x86/coreboot-*/util/me_cleaner/me_cleaner.py 2>&1|head -n1`
+  if [ -z "$MECLEAN" ]; then
+    echo "me_cleaner.py required but not found or specified with -m. Aborting."
+    exit 1;
+  fi
+fi
+
+CAP_ZIP_SHA256SUM="baf7f513227542c507e46735334663f63a0df5be9f6632d7b0f0cca5d3b9f980  P8Z77-M-PRO-ASUS-2203.zip"
+CAP_FILE_SHA256SUM="d9bf292778655d4e20f5db2154cd6a2229e42b60ce670a68d759f1dac757aaf0  P8Z77-M-PRO-ASUS-2203.CAP"
+FINAL_IFD_SHA256SUM="092caeee117de27c0eb30587defcb6449a33c7c325b6f3c47b5a7a79670b5c3f  $BLOBDIR/ifd.bin"
+FINAL_ME_SHA256SUM="8dda1e8360fbb2da05bfcd187f6e7b8a272a67d66bc0074bbfd1410eb35e3e17  $BLOBDIR/me.bin"
+ZIPURL="https://dlcdnets.asus.com/pub/ASUS/mb/LGA1155/P8Z77-M_PRO/P8Z77-M-PRO-ASUS-2203.zip"
+
+ZIPFILENAME=`echo $ZIPURL | sed 's/.*\///'`
+ROMFILENAME=`echo $ZIPFILENAME | sed 's/\.zip$/\.ROM/'`
+
+extractdir=$(mktemp -d)
+echo "### Creating temp dir $extractdir "
+cd "$extractdir"
+
+echo "### Downloading $ZIPURL"
+wget $ZIPURL || { echo "ERROR: wget failed $ZIPURL" && exit 1; }
+echo "### Verifying expected hash of $ZIPFILENAME"
+echo "$CAP_ZIP_SHA256SUM" | sha256sum --check || { echo "Failed sha256sum verification on downloaded binary..." && exit 1; }
+
+echo "### Extracting Archives"
+unzip $ZIPFILENAME || { echo "Failed unzipping $ZIPFILENAME - Tool installed on host?" && exit 1;}
+
+echo "### Verifying expected hash of $ROMFILENAME"
+echo "$CAP_FILE_SHA256SUM" | sha256sum --check || { echo "Failed sha256sum verification on extracted binary..." && exit 1; }
+
+echo "### extracing BIOS from Capsule"
+dd bs=1024 skip=2 if=P8Z77-M-PRO-ASUS-2203.CAP of=P8Z77-M-PRO-ASUS-2203.ROM || { echo "Failed to de-cap the ROM..." && exit 1; }
+
+echo "### Applying me_cleaner to neuter and truncate."
+$MECLEAN -S -r -t -d -O  /tmp/unneeded.bin -D "$BLOBDIR/ifd.bin" -M "$BLOBDIR/me.bin" P8Z77-M-PRO-ASUS-2203.ROM
+
+echo "### Modifying VSCC length and identifiers"
+
+printf '\x00' | dd of="$BLOBDIR/ifd.bin" bs=1 seek=3837 count=1 conv=notrunc
+printf '\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF' | dd of="$BLOBDIR/ifd.bin" bs=1 seek=3568 count=32 conv=notrunc
```

Comment:
Good points, though I would say this has been my daily driver for many months now with no stability issues in this configuration (and the [P8H61 for years before that](https://github.com/osresearch/heads/issues/550)) . The point of the modifications are to result in the removal of the ME from even understanding how to write back to the chip and also keep it in a another [non functioning state ](https://winraid.level1techs.com/t/msi-gp60-2pe-leopard-me-troubles/31656#msg32355)

>It looks like your code clears all entries from the JEDEC and VSCC tables:

Correct, we refer to 0xEFC and 0xEFD for the offset and length, which gives 0xDF0 for 8, giving 4 flash chips and capabilities.  I tend to red in a hex editor as that just what I grew up around. 

> VSCC Table
This section is used to store information to setup flash access for ME. This does not have any effect on the usage of the Flash programming Tool (FPT) 

source: https://github.com/corna/me_cleaner/issues/80#issuecomment-342814683 )

If its a cause for valid concern about some usability or problems that it can creates, certainly lets remove it and revisit it as a separate issue, but I would like to understand the concerns specifically what bad things will happen?  As I said, I've been using for quite some time without any issue - the only thing ive never really tested is flashing with the internal programmer as I use the WP pin anyway. Is that the concern?

Reply:
That's some good research there @rbreslow . Welcome to another rabbit hole ;)

> I had no idea this was a thing. It's starting to make more sense how the VSCC table is part of what allows Intel ME to save state to the SPI flash. I'm not seeing evidence that it's used by anything else.

With so much info out there its hard to tell wheat from chaff sometimes.  There doesn't appear to be much use for this table other than for ME. The main take away for me was that every doc [(even the early stuff](https://web.archive.org/web/20151123141603/http://composter.com.ua/documents/System_Tools_User_Guide_for_Intel_Management_Engine.pdf) such as ME v6) seems to point that removing VSCC stops ME from knowing how to talk to a flash chip. the only question is around the ability for the internal programmer to read and write from/to rom, and a quick test seems to imply an 0xFF VSCC with a zero length doesn't make a difference:

![Screenshot from 2023-04-05 00-03-10](https://user-images.githubusercontent.com/31340423/229937603-41392bc6-8b36-43fe-8277-e56bbe2bbf5d.png)

In terms of why I have always done this the HAP and Neuter/reduce worked well for me but having something else that (n the event something unknown happened) provided an extra layer and took away the ability for ME to write to itself was appealing. 

> Also, I'm curious, what's the output of the MEInfo tool on your machine with and without the VSCC modifications?

Thats a good question and something I never checked as i always relied on HAP and Neuter/Shrink. But with the manufacturer firmware and 0xDF 0x00 (pointer and length)+ 0xFF for VSCC it seems that on this particular machine intelmetool reports that ME is up and running. So, the behaviour in the 8.x versions on asus no longer fails to come up (as per the ME v6 earlier reports)

So as for it being another "soft disable" for ME - tests show that it doesn't look to be the case on this board and the ME version available (i tired ME versions from manufacturer 0802 and 2203 rom capsules)

That said, the documentation you provided (and the older stuff from the me_cleaner thread) does point to removal of this table having implications for the ability for ME to write to the rom chip. Which for the sake of a few printf bytes in a rom extract script, seems like something worth doing?

I guess we will never know for sure if it actually does stop ME -> ROM writes unless someone reverse engineers ME and can trigger an actual ME write event to the SPI, then perform an A/B test with and without this mod.  But we can say that about a lot of ME features, eg no one really knows if even the shrunk code can do 'things' in BUP state. We can just say its 'unlikely' given documentation and understanding. And I think its fair to say from the documentation at hand that removing VSCC (on at least this platform) is likely to have some kind of impact to ME's ability to write to the flash chip.

>How about we leave it as-is and keep using your board for testing? Especially since you've seen months of stability. At the same time, the VSCC stuff is interesting enough that I would like to produce a more thorough analysis and see if it's worth applying to other boards.

sure, sounds good. Im happy to help out as I can.  Its been many years ago - but I think I did this on T430 and X230 as well, though these were test machines mostly for learning, so I cant say that i had them powered up for hours to verify any impact to long term use. But my distant memory tells me I did this to those too.

Reply:
Im going to mark this convo resolved based on discussions and mods as done, but there is certainly something more to look into here for other boards. I will dig out my w530 im pretty sure i did it on that one as well.

Reply:
> I would consider moving the VSCC table modification into a separate issue. We don't do this in any other board, and I'm worried about potential system stability issues.
> 
> It looks like the Z77 chipset is a 7-series chipset. I think I found the datasheet here: [Intel® 7 Series / C216 Chipset Family Platform Controller Hub (PCH)](https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/7-series-chipset-pch-datasheet.pdf). It is very similar to the series 8 chipset, which we can analyze using [ImHex](https://github.com/WerWolv/ImHex) and [the patterns I wrote](https://github.com/rbreslow/ImHex-Patterns/tree/rb/intel-ich8/patterns/intel) for the T440p.
> 
> It looks like your code clears all entries from the JEDEC and VSCC tables:
> Original Flash Descriptor
> Modified Flash Descriptor
> 
> These tables are well documented in [the datasheet](https://web.archive.org/web/20230122164346/https://www.intel.com/content/dam/www/public/us/en/documents/design-guides/i-o-controller-hub-8-9-nvm-map-guide.pdf) for the ICH8 flash descriptor. There are the same number of JEDEC and VSCC entries. The JEDEC entry specifies the SPI vendor/device ID for the component, and the VSCC entry specifies capabilities.

@rbreslow @ThePlexus  idk if this is the right place to ask but I'm trying to get a C216 board added. #1399 Since C216 and Z77 are pretty similar, does that mean I can just copy and paste the exact same command to apply this modification to my ifd.bin too?